example_routing.md

```diff
@@ -0,0 +1,242 @@
+
+# Simple routing spike
+
+Below I've set out some options for simple routing data structuring, and a bunch of other stuff. There two main options
+
+## Option 1
+
+Questions either have a next page when they're unconditional, or a set of conditions if there need to be conditional routing checks. This is not as preferable as the second option, as it requires us to construct more of a branching system whereby all outputs to a conditional question are specified - so all routes out of that question are covered by conditions. The user will implicitly set up routes in different directions, and we'd have to calculate them explicitly. 
+
+There's potentially a bit of value down the line in being extremely explicit about all the ways a route resolves, but it's kind of hard to think about them now.
+
+### api
+```json
+{
+  id: 1,
+  next_page: 2,
+  conditions: nil,
+}
+```
+
+```json
+{
+  id: 1,
+  next_page: nil,
+  conditions: [
+    {question_id: 1, match_answer: "first option", destination_id: 3},
+    {question_id: 1, match_answer: "second option", destination_id: 2},
+  ],
+}
+```
+
+### runner
+
+```ruby
+def calculate_next_page(page)
+  unless page.next_page.nil?
+    return page.next_page
+  end
+
+  page.conditions.each do |condition|
+    given_answer = session[condition.question_id].answer
+    return condition.destination_id if given_answer == condition.match_answer
+  end
+end
+```
+
+## Option 2
+
+Alternatively, question pages will always have a default `next_page` which can be deferred to - but if they have any `conditions`, those will be checked against primarily and take precedence over the `next_page` if any are met. This is what I prefer, as it lets us make minimal additions to questions. 
+
+### api
+
+```json
+{
+  id: 1,
+  next_page: 2,
+  conditions: [
+    {question_id: 1, match_answer: "first option", destination_id: 3},
```

Comment:
Just clarifying the difference between `page.id` and `page.conditions[0].question_id` - is the idea here that:
- `page.id` determines when this condition gets evaluated in the runner
- `page.conditions[0].question_id` determines the question whose value is being interpreted in the condition

So for this `pages` array:
```json
[
  {
    "id": 1,
    "next_page": 2,
    "conditions": []
  },
  {
    "id": 2,
    "next_page": 3,
    "conditions": [{"question_id": 1, "match_answer": "first option", "destination_id": 4}]
  },
  {
    "id": 3,
    "next_page": 4,
    "conditions": []
  },
  {
    "id": 4,
    "next_page": null,
    "conditions": [],
  }
]
```

The user would:
- complete question 1 and get taken to question 2
- complete question 2, and then depending on their answer to question 1 either get taken to question 3 or question 4

Reply:
Yes - because there are conditions in page 2's conditions list, the runner will then check whether the answer to question 1 matches - and if it does, it will use page 4 as the next page